### PR TITLE
body: add knee support

### PR DESCRIPTION
**Pull request by briskspirit**

### CODE REVIEW

**Comment by sshane** [comment]
The body route has FW `b'0.0.01',` for engine, but since that was removed in this PR it doesn't fingerprint and process replay thinks controlsd crashes. If you get a new route it should work

File: selfdrive/car/body/values.py [hunk 1]
```diff
@@ -16,23 +17,35 @@ class CarControllerParams:
 
 class CAR:
   BODY = "COMMA BODY"
+  BODY_KNEE = "COMMA BODY WITH KNEE"
 
 CAR_INFO: Dict[str, CarInfo] = {
   CAR.BODY: CarInfo("comma body", package="All", harness=Harness.none),
+  CAR.BODY_KNEE: CarInfo("comma body + knee", package="All", harness=Harness.none),
```
Comment (adeebshihadeh) [inline selfdrive/car/body/values.py:24]: this is just for docs/harness, don't think we want another entry there
```suggestion
  CAR.BODY_KNEE: None,
```

File: selfdrive/test/process_replay/test_processes.py [hunk 2]
```diff
@@ -38,7 +38,7 @@
 ]
 
 segments = [
-  ("BODY", "regen660D86654BA|2022-07-06--14-27-15--0"),
+  # ("BODY", "regen660D86654BA|2022-07-06--14-27-15--0"),
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/test_processes.py:41]: why was this removed? i'd like to keep process replay coverage on the body

File: selfdrive/car/tests/test_docs.py [hunk 3]
```diff
@@ -59,7 +59,7 @@ def test_year_format(self):
   def test_harnesses(self):
     for car in self.all_cars:
       with self.subTest(car=car):
-        if car.name == "comma body":
+        if car.name in ["comma body", "comma body + knee"]:
```
Comment (adeebshihadeh) [inline selfdrive/car/tests/test_docs.py:62]: this should use carParams.notCar

File: selfdrive/car/tests/routes.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
   HYUNDAI.ELANTRA_GT_I30,
   HYUNDAI.GENESIS_G90,
   HYUNDAI.KIA_OPTIMA_H,
+  COMMA.BODY_KNEE,
```
Comment (adeebshihadeh) [inline selfdrive/car/tests/routes.py:27]: let's add a test route

File: selfdrive/car/body/values.py [hunk 5]
```diff
@@ -16,23 +17,35 @@ class CarControllerParams:
...
 }
 
 FW_VERSIONS = {
   CAR.BODY: {
     (Ecu.engine, 0x720, None): [
-      b'0.0.01',
-      b'02/27/2022'
+      # b'0.0.02',
+      b'ELECTRIC0'
```
Comment (adeebshihadeh) [inline selfdrive/car/body/values.py:31]: I think we'll want to finish auto fw update before merging, or at least before release

File: selfdrive/car/body/values.py [hunk 6]
```diff
@@ -16,23 +17,35 @@ class CarControllerParams:
...
+      b'1fc01d70' # git hash of the firmware used
+    ],
+  },
+  CAR.BODY_KNEE: {
+    (Ecu.engine, 0x720, None): [
+      # b'0.0.02',
+      b'ELECTRIC1'
+    ],
+    (Ecu.debug, 0x721, None): [
+      b'1fc01d70' # git hash of the firmware used
```
Comment (adeebshihadeh) [inline selfdrive/car/body/values.py:43]: both main body and knee ecu should report their git hash

File: selfdrive/test/process_replay/test_processes.py (hunk 2 above)
Comment (adeebshihadeh) [reply selfdrive/test/process_replay/test_processes.py:41]: let's also use a knee route for more coverage

File: selfdrive/car/body/carcontroller.py [hunk 7]
```diff
@@ -54,8 +72,11 @@ def update(self, CC, CS):
                            (speed_error > 0 and self.speed_pid.error_integral >= MAX_POS_INTEGRATOR))
       angle_setpoint = self.speed_pid.update(speed_error, freeze_integrator=freeze_integrator)
 
-      # Clip angle error, this is enough to get up from stands
-      angle_error = np.clip((-CC.orientationNED[1]) - angle_setpoint, -MAX_ANGLE_ERROR, MAX_ANGLE_ERROR)
+      if self.CP.carFingerprint == CAR.BODY_KNEE:
+        knee_error = (np.radians(CS.knee_angle_r - CS.knee_angle_r))
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:76]:
```suggestion
        knee_error = np.radians(CS.knee_angle_r - CS.knee_angle_r)
```

File: selfdrive/car/body/carcontroller.py [hunk 8]
```diff
@@ -66,25 +87,68 @@ def update(self, CC, CS):
...
       torque_l = int(np.clip(self.torque_l_filtered, -MAX_TORQUE, MAX_TORQUE))
+      torque_r = int(np.clip(self.torque_r_filtered, -MAX_TORQUE, MAX_TORQUE))
+
+    # Knee angle controls
+    if self.CP.carFingerprint == CAR.BODY_KNEE and CC.enabled:
+      knee_angle_desired_left = CC.actuators.gas
+      knee_angle_desired_right = CC.actuators.brake
+
+      knee_angle_measured_left = CS.knee_angle_l
+      knee_angle_measured_right = CS.knee_angle_r
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:110]: these should be in a cereal struct

File: selfdrive/car/body/carcontroller.py [hunk 9]
```diff
@@ -66,25 +87,68 @@ def update(self, CC, CS):
...
+                                        self.torque_knee_l_filtered + MAX_KNEE_TORQUE_RATE)
+      self.torque_knee_r_filtered = np.clip(knee_torque_r,
+                                        self.torque_knee_r_filtered - MAX_KNEE_TORQUE_RATE,
+                                        self.torque_knee_r_filtered + MAX_KNEE_TORQUE_RATE)
+
+      knee_torque_l = int(np.clip(self.torque_knee_l_filtered, -MAX_KNEE_TORQUE_LEFT, MAX_KNEE_TORQUE_LEFT))
+      knee_torque_r = int(np.clip(self.torque_knee_r_filtered, -MAX_KNEE_TORQUE_RIGHT, MAX_KNEE_TORQUE_RIGHT))
+
+      # Extra safety for when training stands attached:
+      if knee_angle_measured_left<=30 and knee_torque_l > 0:
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:135]:
```suggestion
      if knee_angle_measured_left <= 30 and knee_torque_l > 0:
```

File: selfdrive/car/body/carcontroller.py [hunk 10]
```diff
@@ -66,25 +87,68 @@ def update(self, CC, CS):
...
+
+      freeze_integrator = ((knee_angle_error_left < 0 and self.knee_angle_left_pid.error_integral <= -MAX_KNEE_ANGLE_INTEGRATOR_LEFT) or
+                           (knee_angle_error_left > 0 and self.knee_angle_left_pid.error_integral >= MAX_KNEE_ANGLE_INTEGRATOR_LEFT))
+      knee_torque_l = self.knee_angle_left_pid.update(knee_angle_error_left, freeze_integrator=freeze_integrator)
+
+      freeze_integrator = ((knee_angle_error_right < 0 and self.knee_angle_right_pid.error_integral <= -MAX_KNEE_ANGLE_INTEGRATOR_RIGHT) or
+                           (knee_angle_error_right > 0 and self.knee_angle_right_pid.error_integral >= MAX_KNEE_ANGLE_INTEGRATOR_RIGHT))
+      knee_torque_r = self.knee_angle_right_pid.update(knee_angle_error_right, freeze_integrator=freeze_integrator)
+
+    # Torque rate limits
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:123]:
```suggestion
      # Torque rate limits
```

File: selfdrive/car/body/carcontroller.py [hunk 11]
```diff
@@ -66,25 +87,68 @@ def update(self, CC, CS):
...
+      knee_torque_l = int(np.clip(self.torque_knee_l_filtered, -MAX_KNEE_TORQUE_LEFT, MAX_KNEE_TORQUE_LEFT))
+      knee_torque_r = int(np.clip(self.torque_knee_r_filtered, -MAX_KNEE_TORQUE_RIGHT, MAX_KNEE_TORQUE_RIGHT))
+
+      # Extra safety for when training stands attached:
+      if knee_angle_measured_left<=30 and knee_torque_l > 0:
+        knee_torque_l = 0
+      elif knee_angle_measured_left >= 330 and knee_torque_l < 0:
+        knee_torque_l = 0
+
+      # Do not try to balance if knee and hip have excessive angle on start
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:140]: is the rate limit low enough to safely fade in control if a body is already onroad and the user moves it into position?

File: selfdrive/car/docs_definitions.py [hunk 12]
```diff
@@ -162,7 +162,7 @@ def get_detail_sentence(self, CP):
       return sentence_builder.format(car_model=f"{self.make} {self.model}", alc=alc, acc=acc)
 
     else:
-      if CP.carFingerprint == "COMMA BODY":
+      if CP.carFingerprint in ["COMMA BODY", "COMMA BODY WITH KNEE"]:
```
Comment (adeebshihadeh) [inline selfdrive/car/docs_definitions.py:165]: shouldn't be necessary since knee CarInfo is none

File: selfdrive/car/body/carcontroller.py [hunk 13]
```diff
@@ -3,29 +3,45 @@
...
+MAX_TORQUE = 700
 MAX_TORQUE_RATE = 50
 MAX_ANGLE_ERROR = np.radians(7)
-MAX_POS_INTEGRATOR = 0.2   # meters
-MAX_TURN_INTEGRATOR = 0.1  # meters
+MAX_POS_INTEGRATOR = 10.0   # meters
+MAX_TURN_INTEGRATOR = 10.0  # meters
+
+MAX_KNEE_TORQUE_LEFT = 400 # knee motor
+MAX_KNEE_TORQUE_RIGHT = 200 # hip motor
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:17]: call these knee and hip instead? using left/right seems confusing here. left/right seems like it's just an implementation detail?

File: selfdrive/car/body/carcontroller.py [hunk 14]
```diff
@@ -54,8 +72,11 @@ def update(self, CC, CS):
                            (speed_error > 0 and self.speed_pid.error_integral >= MAX_POS_INTEGRATOR))
       angle_setpoint = self.speed_pid.update(speed_error, freeze_integrator=freeze_integrator)
 
-      # Clip angle error, this is enough to get up from stands
-      angle_error = np.clip((-CC.orientationNED[1]) - angle_setpoint, -MAX_ANGLE_ERROR, MAX_ANGLE_ERROR)
+      if self.CP.carFingerprint == CAR.BODY_KNEE:
+        knee_error = (np.radians(CS.knee_angle_r - CS.knee_angle_r))
+        knee_error = 0 # Ignore knee angle error for now
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carcontroller.py:77]: remove the line above then?

File: selfdrive/car/body/carcontroller.py (hunk 11 above)
Comment (briskspirit) [reply selfdrive/car/body/carcontroller.py:140]: What do you mean by "moves it into position" ? moves body or knee?

File: selfdrive/car/body/carcontroller.py (hunk 8 above)
Comment (briskspirit) [reply selfdrive/car/body/carcontroller.py]: in car state? say as SensorAngles struct?

File: selfdrive/car/docs_definitions.py [hunk 15]
```diff
@@ -162,7 +162,7 @@ def get_detail_sentence(self, CP):
       return sentence_builder.format(car_model=f"{self.make} {self.model}", alc=alc, acc=acc)
 
     else:
-      if CP.carFingerprint in ["COMMA BODY", "COMMA BODY WITH KNEE"]:
+      if CP.carFingerprint == "COMMA BODY":
```
Comment (sshane) [inline selfdrive/car/docs_definitions.py:165]: This can be notCar, now that we're not showing the body w/ knee as a new car

File: selfdrive/car/docs_definitions.py [hunk 16]
```diff
@@ -162,10 +162,7 @@ def get_detail_sentence(self, CP):
       return sentence_builder.format(car_model=f"{self.make} {self.model}", alc=alc, acc=acc)
 
     else:
-      if CP.carFingerprint == "COMMA BODY":
-        return "The body is a robotics dev kit that can run openpilot. <a href='https://www.commabody.com'>Learn more.</a>"
-      else:
-        raise Exception(f"This notCar does not have a detail sentence: {CP.carFingerprint}")
+      return "The body is a robotics dev kit that can run openpilot. <a href='https://www.commabody.com'>Learn more.</a>"
```
Comment (adeebshihadeh) [inline selfdrive/car/docs_definitions.py:165]: revert the changes here, we want to support lots of other not cars :)

File: selfdrive/car/body/carstate.py [hunk 17]
```diff
@@ -57,4 +59,19 @@ def get_can_parser(CP):
       ("BODY_DATA", 1),
     ]
 
+    if CP.carFingerprint == CAR.BODY_KNEE:
+      signals += [
+        ("SPEED_L", "KNEE_MOTORS_DATA"),
+        ("SPEED_R", "KNEE_MOTORS_DATA"),
+        ("COUNTER", "KNEE_MOTORS_DATA"),
+        ("CHECKSUM", "KNEE_MOTORS_DATA"),
```
Comment (adeebshihadeh) [inline selfdrive/car/body/carstate.py:67]: what do you use counter and checksum for? they don't need to be added here to validate them in the can parser

**Comment by adeebshihadeh** [comment]
Gonna close this for now and keep the branch around
